File: jurij_veresciaka/U2/code/php/includes/helpers/TimeHelper.php
<?php

class TimeHelper
{
    public function getNaktinisLaikas($date_start, $date_end, $night_start, $night_end)
    {
        ///////////////////////////////////////////////
        //CHECK TYPES                                //
        ///////////////////////////////////////////////

        if (!($date_start instanceof DateTime)
        ) {
            throw new Exception('date_start wrong parameter');
        }

        if (!($date_end instanceof DateTime)
        ) {
            throw new Exception('date_end wrong parameter');
        }

        if (!($night_start instanceof DateTime)
        ) {
            throw new Exception('night_start wrong parameter');
        }

        if (!($night_end instanceof DateTime)
        ) {
            throw new Exception('night_end wrong parameter');
        }

        ///////////////////////////////////////////////
        //CHECK VALUES                               //
        ///////////////////////////////////////////////

        if (!($date_start < $date_end)
        ) {
            throw new Exception('!(date_start < date_end)');
        }

        if (!($night_start < $night_end)
        ) {
            throw new Exception('!(night_start < night_end)');
        }

        ///////////////////////////////////////////////
        //LOGIC                                      //
        ///////////////////////////////////////////////

        $total_days = $date_start->diff($date_end)->days;

        $is_days_the_same = FALSE;
        if ($date_start->format('Y-m-d') == $date_end->format('Y-m-d')) {
            $is_days_the_same = TRUE;
        }

        $night_minutes = 0;

        if ($is_days_the_same) {
            $night_minutes += $this->getNightMinutesPerDay($date_start, $date_end, $night_start, $night_end);
        } else {
            $night_minutes += $this->getNightMinutesPerDay(
                $date_start,
                new DateTime($date_start->format('Y-m-d') . ' ' . '23:59:00'),
                $night_start,
                $night_end
            );
            $night_minutes += $this->getNightMinutesPerDay(
                new DateTime($date_end->format('Y-m-d') . ' ' . '00:00:00'),
                $date_end,
                $night_start,
                $night_end
            );

            $night_total = $this->getMinutesFromTime($night_end) - $this->getMinutesFromTime($night_start);

            if ($total_days > 0) {
                if ($this->getMinutesFromTime($date_end) >= $this->getMinutesFromTime($date_start)) {
                    $night_minutes += ($total_days - 1) * $night_total;
                } else {
                    $night_minutes += $total_days * $night_total;
                }
            }
        }

        return $night_minutes;
    }

    ///////////////////////////////////////////////
    //HELPERS                                    //
    ///////////////////////////////////////////////

    public function getNightMinutesPerDay($day_start, $day_end, $night_start, $night_end)
    {
        ///////////////////////////////////////////////
        //CHECK TYPES                                //
        ///////////////////////////////////////////////

        if (!($day_start instanceof DateTime)
        ) {
            throw new Exception('day_start wrong parameter');
        }

        if (!($day_end instanceof DateTime)
        ) {
            throw new Exception('day_end wrong parameter');
        }

        if (!($night_start instanceof DateTime)
        ) {
            throw new Exception('night_start wrong parameter');
        }

        if (!($night_end instanceof DateTime)
        ) {
            throw new Exception('night_end wrong parameter');
        }

        ///////////////////////////////////////////////
        //CHECK VALUES                               //
        ///////////////////////////////////////////////

        if (!($day_start <= $day_end)
        ) {
            throw new Exception('!(day_start <= day_end)');
        }

        if (!($night_start < $night_end)
        ) {
            throw new Exception('!(night_start < night_end)');
        }

        ///////////////////////////////////////////////
        //LOGIC                                      //
        ///////////////////////////////////////////////

        $day_start_minutes = $this->getMinutesFromTime($day_start);
        $day_end_minutes = $this->getMinutesFromTime($day_end);
        $night_start_minutes = $this->getMinutesFromTime($night_start);
        $night_end_minutes = $this->getMinutesFromTime($night_end);

        $day_total_minutes = $day_end_minutes - $day_start_minutes;
        $night_total_minutes = $night_end_minutes - $night_start_minutes;

        // day   /--/     ||      /--/
        // night     /--/ || /--/
        if (($day_end_minutes <= $night_start_minutes) || ($day_start_minutes >= $night_end_minutes)) {
            return 0;
        }
        // day   /-----/
        // night  /--/
        if (($day_start_minutes <= $night_start_minutes) && ($day_end_minutes >= $night_end_minutes)) {
            return $night_total_minutes;
        }
        // day     /--/
        // night /-----/
        if (($day_start_minutes >= $night_start_minutes) && ($day_end_minutes <= $night_end_minutes)) {
            return $day_total_minutes;
        }
        // day    /-----/    || /-----/
        // night     /-----/ || /--------/
        if (($day_start_minutes <= $night_start_minutes) && ($day_end_minutes < $night_end_minutes)) {
            return ($day_end_minutes - $night_start_minutes);
        }
        // day       /-----/ ||    /-----/
        // night  /-----/    || /--------/
        if (($day_start_minutes > $night_start_minutes) && ($day_end_minutes >= $night_end_minutes)) {
            return ($night_end_minutes - $day_start_minutes);
        }
    }

    public function getMinutesFromTime($time)
    {
        ///////////////////////////////////////////////
        //CHECK TYPES                                //
        ///////////////////////////////////////////////

        if (!($time instanceof DateTime)
        ) {
            throw new Exception('time wrong parameter');
        }

        ///////////////////////////////////////////////
        //LOGIC                                      //
        ///////////////////////////////////////////////

        $hours = (int)$time->format('H');
        $minutes = (int)$time->format('i');

        $minutes = $hours * 60 + $minutes;

        return $minutes;
    }
}

File: jurij_veresciaka/U2/code/csharp/csharp/TimeHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace csharp
{
    public class TimeHelper
    {
        public static int getNaktinisLaikas(DateTime dateStart, DateTime dateEnd, DateTime nightStart, DateTime nightEnd)
        {
            ///////////////////////////////////////////////
            //CHECK VALUES                               //
            ///////////////////////////////////////////////

            if (DateTime.Compare(dateStart, dateEnd) >= 0)
            {
                throw new System.ArgumentException("!(dateStart < dateEnd)", "dateStart, dateEnd");
            }

            if (DateTime.Compare(nightStart, nightEnd) >= 0)
            {
                throw new System.ArgumentException("!(nightStart < nightEnd)", "nightStart, nightEnd");
            }

            ///////////////////////////////////////////////
            //LOGIC                                      //
            ///////////////////////////////////////////////

            System.TimeSpan totalDays = dateEnd.Subtract(dateStart);

            bool isDaysTheSame = false;

            if (DateTime.Compare(
                new DateTime(dateStart.Year, dateStart.Month, dateStart.Day),
                new DateTime(dateEnd.Year, dateEnd.Month, dateEnd.Day)
                ) == 0)
            {
                isDaysTheSame = true;
            }

            int nightMinutes = 0;

            if (isDaysTheSame) {
                nightMinutes += getNightMinutesPerDay(dateStart, dateEnd, nightStart, nightEnd);
            }
            else
            {
                nightMinutes += getNightMinutesPerDay(dateStart, new DateTime(dateEnd.Year, dateEnd.Month, dateEnd.Day, 23, 59, 0), nightStart, nightEnd);
                nightMinutes += getNightMinutesPerDay(new DateTime(dateStart.Year, dateStart.Month, dateStart.Day, 0, 0, 0), dateEnd, nightStart, nightEnd);
            }

            int nightTotal = getMinutesFromTime(nightEnd) - getMinutesFromTime(nightStart);

            if (totalDays.Days > 0) {
                if (getMinutesFromTime(dateEnd) >= getMinutesFromTime(dateStart)) {
                    nightMinutes += (totalDays.Days - 1) * nightTotal;
                } else {
                    nightMinutes += totalDays.Days * nightTotal;
                }
            }

            return nightMinutes;
        }

        ///////////////////////////////////////////////
        //HELPERS                                    //
        ///////////////////////////////////////////////

        public static int getNightMinutesPerDay(DateTime dayStart, DateTime dayEnd, DateTime nightStart, DateTime nightEnd)
        {
            ///////////////////////////////////////////////
            //CHECK VALUES                               //
            ///////////////////////////////////////////////

            if (DateTime.Compare(dayStart, dayEnd) > 0)
            {
                throw new System.ArgumentException("!(dayStart <= dayStart)", "dayStart, dayEnd");
            }

            if (DateTime.Compare(nightStart, nightEnd) >= 0)
            {
                throw new System.ArgumentException("!(nightStart < nightEnd)", "nightStart, nightEnd");
            }

            ///////////////////////////////////////////////
            //LOGIC                                      //
            ///////////////////////////////////////////////

            int dayStartMinutes = getMinutesFromTime(dayStart);
            int dayEndMinutes = getMinutesFromTime(dayEnd);
            int nightStartMinutes = getMinutesFromTime(nightStart);
            int nightEndMinutes = getMinutesFromTime(nightEnd);

            int dayTotalMinutes = dayEndMinutes - dayStartMinutes;
            int nightTotalMinutes = nightEndMinutes - nightStartMinutes;

            // day   /--/     ||      /--/
            // night     /--/ || /--/
            if ((dayEndMinutes <= nightStartMinutes) || (dayStartMinutes >= nightEndMinutes)) {
                return 0;
            }
            // day   /-----/
            // night  /--/
            if ((dayStartMinutes <= nightStartMinutes) && (dayEndMinutes >= nightEndMinutes)) {
                return nightTotalMinutes;
            }
            // day     /--/
            // night /-----/
            if ((dayStartMinutes >= nightStartMinutes) && (dayEndMinutes <= nightEndMinutes)) {
                return dayTotalMinutes;
            }
            // day    /-----/    || /-----/
            // night     /-----/ || /--------/
            if ((dayStartMinutes <= nightStartMinutes) && (dayEndMinutes < nightEndMinutes)) {
                return dayEndMinutes - nightStartMinutes;
            }
            // day       /-----/ ||    /-----/
            // night  /-----/    || /--------/
            if ((dayStartMinutes > nightStartMinutes) && (dayEndMinutes >= nightEndMinutes)) {
                return nightEndMinutes - dayStartMinutes;
            }

            return 0;
        }


        private static int getMinutesFromTime(DateTime time)
        {
            ///////////////////////////////////////////////
            //LOGIC                                      //
            ///////////////////////////////////////////////

            int hours = time.Hour;
            int minutes = time.Minute;

            minutes = hours * 60 + minutes;
        
            return minutes;
        }
    }
}
